### PR TITLE
Final project - Back end

### PR DESCRIPTION
**Pull request by sheylaPozo**
##### Front end PR
----
## Final project - Back end
**Database:**

- Usage: PostgreSql.
- Created tables: user, flats, reservations, cities.

_Models:_

- All associations are added to the models, so the relations made into the database between the tables, can be used in rails as well.
- All validations are added to the models, in order to avoid incorrect "inputs values" from a user into the database.

_Controllers:_
**Users:**

-:create is used to add a new object of user to the database.
- :destroy is used to delete a user from the database.
- private method that ensures that a specific parameter is present and avoid penetration into the project. As someone from outside to get an id of user and get personal information.

**Flats:**

:index is used to list all the flats from the database.
:create is used to add a new object of flat to the database.
:destroy is used to delete a flat from the database.
private method that ensures that a specific parameter is present and avoid penetration into the project. As someone from outside to get an id of user and get personal information.

**Reservations:**

:index is used to list all the reservations that a current User has.
:create is used to add a new object of reservation to the database.
:destroy is used to delete a reservation from the database.
private method that ensures that a specific parameter is present and avoid penetration into the project. As someone from outside to get an id of user and get personal information.

**Routes:**

The routes are created for each method in the controllers, in order the to create endpoints of the Api.

**API Documentation:**

API Documentation is created, so the Api can be easily used by users.

**_Contributors:_**
@2bleO

Thank you for taking the time to review our PR! 😃😁📌

https://finalcapstoneflats.herokuapp.com/ there was an error on the rake file not found that didn't allow the deployment to go through, and we needed the heroku deployment site for the API end points connection.

### CODE REVIEW

File: app/controllers/api/v1/cities_controller.rb [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+          render json: { message: 'City created successfully', data: @city }, status: :created
+        else
+          render json: { message: 'There was an error' }, status: :unauthorized
+        end
+      end
+
+      def destroy
+        city = City.find(params[:id])
+        if city
+          city.reservations&.delete_all
```
Comment (MuriloRoque) [inline app/controllers/api/v1/cities_controller.rb:27]: - [ ] Take a look [here](https://guides.rubyonrails.org/association_basics.html#options-for-belongs-to-dependent), you can use `dependent` in `has_many` to automatically destroy all reservations, wouldn't that be better?

File: config/routes.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  mount Rswag::Api::Engine => '/api-docs'
+  namespace :api do
+    namespace :v1 do
+      # authentication routes
+      post :signin, to: 'auth#signin'
+      post :signup, to: 'auth#signup'
+      post :signout, to: 'auth#signout' 
+      
+      # cars
+      resources :flats
```
Comment (MuriloRoque) [inline config/routes.rb:13]: - [ ] This creates all routes, including unused ones, I would select only the ones I'm using.

File: config/routes.rb [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+      post "delete_flat", to: "flats#delete_flat"
+
+      # reservations
+      get "reservations/:user_id", to: "reservations#index"
+      resources :reservations, only: [:create, :destroy]
+
+      # cities
+      resources :cities
```
Comment (MuriloRoque) [inline config/routes.rb:21]: - [ ] This creates all routes, including unused ones, I would select only the ones I'm using.

File: app/controllers/pages_controller.rb [hunk 4]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+class PagesController < ActionController::Base
+  def index
+    redirect_to '/api-docs'
+  end
+end
```
Comment (MuriloRoque) [inline app/controllers/pages_controller.rb:7]: - [ ] Are you using this controller? If not, it's better to delete it.

**Review by MuriloRoque — CHANGES_REQUESTED** [review]
Hi @sheylaPozo, @2bleO, 

Good job so far!
There are some issues that you still need to work on to prepare your project for the final evaluation but you are almost there!

## Suggested changes
_Check the comments under the review._

You can use as many of my suggestions as you want. If there is anything you would like to skip - feel free to do that. However, I strongly recommend you to take them into account as they can make your code better.

You can also consider:

- [ ] Adding some tests would be great.

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

**Comment by sheylaPozo** [comment]
thank youu !!

**Pull request merged**